File: validator/auth/cells.cpp
#include <sodium.h>

#include "vm/boc.h"
#include "vm/cells/CellBuilder.h"
#include "vm/cells/CellSlice.h"

#include "cells.h"
namespace tos::auth {
namespace {
td::Slice slice(std::span<const std::uint8_t> b) {
  return td::Slice(reinterpret_cast<const char*>(b.data()), b.size());
}
std::size_t capacity(std::size_t size) {
  std::size_t n = 120;
  while (size > 4 * n)
    n *= 4;
  return n;
}
td::Ref<vm::Cell> pack_node(std::span<const std::uint8_t> raw) {
  vm::CellBuilder b;
  if (raw.size() <= 120)
    b.store_long(0, 1).store_long(raw.size(), 7).store_bytes(slice(raw));
  else {
    auto cap = capacity(raw.size());
    b.store_long(1, 1).store_long((raw.size() + cap - 1) / cap, 3).store_long(raw.size(), 32);
    for (std::size_t offset = 0; offset < raw.size(); offset += cap)
      b.store_ref(pack_node(raw.subspan(offset, std::min(cap, raw.size() - offset))));
  }
  return b.finalize();
}
bool unpack_node(td::Ref<vm::Cell> cell, std::size_t expected, unsigned depth, unsigned& occurrences, Bytes& out) {
  if (cell.is_null() || depth > 10 || ++occurrences > 400000 || cell->get_level() != 0)
    return false;
  vm::CellSlice s{vm::NoVm{}, cell};
  if (!s.is_valid() || s.is_special() || s.size() < 1)
    return false;
  auto branch = s.fetch_ulong(1);
  if (expected <= 120) {
    if (branch != 0 || s.size() < 7)
      return false;
    auto n = s.fetch_ulong(7);
    if (n != expected || n == 0 || s.size() != 8 * n || s.size_refs() != 0)
      return false;
    std::array<char, 120> bytes{};
    if (!s.fetch_bytes(td::MutableSlice(bytes.data(), n)))
      return false;
    out.insert(out.end(), bytes.begin(), bytes.begin() + n);
    return true;
  }
  if (branch != 1 || s.size() != 35)
    return false;
  auto n = s.fetch_ulong(3), length = s.fetch_ulong(32);
  auto cap = capacity(expected);
  if (length != expected || n != (expected + cap - 1) / cap || n < 2 || n > 4 || s.size_refs() != n)
    return false;
  for (std::size_t offset = 0; offset < expected; offset += cap)
    if (!unpack_node(s.fetch_ref(), std::min(cap, expected - offset), depth + 1, occurrences, out))
      return false;
  return true;
}
}  // namespace
Result<td::Ref<vm::Cell>> pack_bytes(std::span<const std::uint8_t> bytes) {
  if (bytes.empty() || bytes.size() > max_object_bytes)
    return Error{"object-bound"};
  Hash hash;
  if (crypto_hash_sha256(hash.data(), bytes.data(), bytes.size()) != 0)
    return Error{"backend-error"};
  vm::CellBuilder b;
  b.store_long(0x76616231, 32)
      .store_long(1, 16)
      .store_long(bytes.size(), 32)
      .store_bytes(slice(hash))
      .store_ref(pack_node(bytes));
  return td::Ref<vm::Cell>(b.finalize());
}
Result<Bytes> unpack_bytes(td::Ref<vm::Cell> cell) {
  if (cell.is_null() || cell->get_level() != 0)
    return Error{"cell-level"};
  vm::CellSlice s{vm::NoVm{}, cell};
  if (!s.is_valid() || s.is_special() || s.size() != 336 || s.size_refs() != 1)
    return Error{"auth-bytes-shape"};
  if (s.fetch_ulong(32) != 0x76616231 || s.fetch_ulong(16) != 1)
    return Error{"auth-bytes-version"};
  auto length = s.fetch_ulong(32);
  if (length == 0 || length > max_object_bytes)
    return Error{"object-bound"};
  Hash expected;
  if (!s.fetch_bytes(td::MutableSlice(reinterpret_cast<char*>(expected.data()), 32)))
    return Error{"auth-bytes-hash"};
  Bytes raw;
  raw.reserve(length);
  unsigned occurrences = 0;
  if (!unpack_node(s.fetch_ref(), length, 0, occurrences, raw))
    return Error{"tree-canonical"};
  Hash actual;
  if (crypto_hash_sha256(actual.data(), raw.data(), raw.size()) != 0)
    return Error{"backend-error"};
  if (actual != expected)
    return Error{"auth-bytes-hash"};
  return raw;
}
Result<Bytes> serialize_bytes(std::span<const std::uint8_t> bytes) {
  auto root = pack_bytes(bytes);
  if (!root.ok())
    return root.error();
  auto boc = vm::std_boc_serialize(root.value(), 31);
  if (boc.is_error())
    return Error{"boc-serialize"};
  auto value = boc.move_as_ok();
  if (value.size() > 67108864)
    return Error{"boc-bound"};
  return Bytes(value.as_slice().ubegin(), value.as_slice().uend());
}
Result<Bytes> deserialize_bytes(std::span<const std::uint8_t> boc) {
  if (boc.empty() || boc.size() > 67108864)
    return Error{"boc-bound"};
  vm::BagOfCells::Info info;
  auto size = info.parse_serialized_header(slice(boc).substr(0, std::min<std::size_t>(256, boc.size())));
  if (size <= 0 || static_cast<std::size_t>(size) != boc.size() || info.root_count != 1 || info.cell_count <= 0 ||
      info.cell_count > 400001 || info.absent_count != 0)
    return Error{"boc-header-bound"};
  vm::BagOfCells parsed;
  auto consumed = parsed.deserialize(slice(boc), 1);
  if (consumed.is_error())
    return Error{"boc-decode"};
  if (consumed.ok() != static_cast<long long>(boc.size()) || parsed.get_root_count() != 1)
    return Error{"boc-trailing"};
  return unpack_bytes(parsed.get_root_cell());
}
}  // namespace tos::auth

File: tosctl/src/validator-auth-native/src/cells.rs
use crate::native;
use chain_block::{BocReader, BuilderData, Cell, CellType, IBitstring, SliceData};
use sha2::{Digest, Sha256};
use std::{collections::HashSet, io::Cursor};
use tos_validator_auth::codec::Error;
pub const MAX_OBJECT: usize = 33_554_432;
pub const MAX_BOC: usize = 67_108_864;
const MAX_CELLS: usize = 400_001;
fn capacity(size: usize) -> usize {
    let mut result = 120;
    while size > 4 * result {
        result *= 4;
    }
    result
}
fn pack_node(raw: &[u8]) -> Result<Cell, Error> {
    let mut builder = BuilderData::new();
    if raw.len() <= 120 {
        native(builder.append_bit_zero())?;
        native(builder.append_bits(raw.len(), 7))?;
        native(builder.append_raw(raw, raw.len() * 8))?;
    } else {
        let cap = capacity(raw.len());
        native(builder.append_bit_one())?;
        native(builder.append_bits(raw.len().div_ceil(cap), 3))?;
        native(builder.append_u32(raw.len() as u32))?;
        for chunk in raw.chunks(cap) {
            native(builder.checked_append_reference(pack_node(chunk)?))?;
        }
    }
    native(builder.into_cell())
}
pub fn pack_bytes(raw: &[u8]) -> Result<Cell, Error> {
    if raw.is_empty() || raw.len() > MAX_OBJECT {
        return Err(Error("object-bound"));
    }
    let mut builder = BuilderData::new();
    native(builder.append_u32(0x76616231))?;
    native(builder.append_u16(1))?;
    native(builder.append_u32(raw.len() as u32))?;
    native(builder.append_raw(&Sha256::digest(raw), 256))?;
    native(builder.checked_append_reference(pack_node(raw)?))?;
    native(builder.into_cell())
}
fn unpack_node(
    cell: Cell,
    expected: usize,
    depth: u8,
    occurrences: &mut usize,
    out: &mut Vec<u8>,
) -> Result<(), Error> {
    *occurrences = occurrences.checked_add(1).ok_or(Error("tree-canonical"))?;
    if depth > 10
        || *occurrences > 400_000
        || cell.level() != 0
        || cell.cell_type() != CellType::Ordinary
    {
        return Err(Error("tree-canonical"));
    }
    let mut slice = native(SliceData::load_cell(cell))?;
    let branch = native(slice.get_next_bit())?;
    if expected <= 120 {
        let n = native(slice.get_next_int(7))? as usize;
        if branch
            || n != expected
            || n == 0
            || slice.remaining_bits() != n * 8
            || slice.remaining_references() != 0
        {
            return Err(Error("tree-canonical"));
        }
        out.extend_from_slice(&native(slice.get_next_bits(n * 8))?);
    } else {
        if !branch || slice.remaining_bits() != 35 {
            return Err(Error("tree-canonical"));
        }
        let n = native(slice.get_next_int(3))? as usize;
        let length = native(slice.get_next_u32())? as usize;
        let cap = capacity(expected);
        if length != expected
            || n != expected.div_ceil(cap)
            || !(2..=4).contains(&n)
            || slice.remaining_references() != n
        {
            return Err(Error("tree-canonical"));
        }
        for offset in (0..expected).step_by(cap) {
            unpack_node(
                native(slice.checked_drain_reference())?,
                cap.min(expected - offset),
                depth + 1,
                occurrences,
                out,
            )?;
        }
    }
    Ok(())
}
pub fn unpack_bytes(cell: Cell, remaining: usize) -> Result<Vec<u8>, Error> {
    if cell.level() != 0 || cell.cell_type() != CellType::Ordinary {
        return Err(Error("cell-level"));
    }
    let mut slice = native(SliceData::load_cell(cell))?;
    if slice.remaining_bits() != 336 || slice.remaining_references() != 1 {
        return Err(Error("auth-bytes-shape"));
    }
    if native(slice.get_next_u32())? != 0x76616231 || native(slice.get_next_u16())? != 1 {
        return Err(Error("auth-bytes-version"));
    }
    let length = native(slice.get_next_u32())? as usize;
    if length == 0 || length > MAX_OBJECT {
        return Err(Error("object-bound"));
    }
    if length > remaining {
        return Err(Error("state-resource"));
    }
    let expected = native(slice.get_next_bits(256))?;
    let mut raw = Vec::with_capacity(length);
    unpack_node(native(slice.checked_drain_reference())?, length, 0, &mut 0, &mut raw)?;
    if Sha256::digest(&raw).as_slice() != expected {
        return Err(Error("auth-bytes-hash"));
    }
    Ok(raw)
}
/// Inspect declared sizes before the native reader allocates its cell/index tables.
pub fn read_boc(bytes: &[u8], all_reachable: bool) -> Result<Cell, Error> {
    if bytes.is_empty() || bytes.len() > MAX_BOC {
        return Err(Error("boc-bound"));
    }
    let mut cursor = Cursor::new(bytes);
    let mut reader = BocReader::new().set_max_cell_depth(1024);
    let (header, _) = native(reader.read_header(&mut cursor))?;
    if header.roots_count != 1 || header.cells_count == 0 || header.cells_count > MAX_CELLS {
        return Err(Error("boc-header-bound"));
    }
    let index = if header.index_included {
        header.cells_count.checked_mul(header.offset_size).ok_or(Error("boc-header-bound"))?
    } else {
        0
    };
    let size = (cursor.position() as usize)
        .checked_add(index)
        .and_then(|n| n.checked_add(header.tot_cells_size))
        .and_then(|n| n.checked_add(if header.has_crc { 4 } else { 0 }))
        .ok_or(Error("boc-header-bound"))?;
    if size != bytes.len() {
        return Err(Error("boc-header-bound"));
    }
    cursor.set_position(0);
    let parsed = native(reader.read(&mut cursor))?;
    if cursor.position() != bytes.len() as u64 {
        return Err(Error("boc-trailing"));
    }
    let root = native(parsed.withdraw_single_root())?;
    if all_reachable {
        let mut seen = HashSet::new();
        let mut pending = vec![root.clone()];
        while let Some(cell) = pending.pop() {
            if !seen.insert(cell.repr_hash()) {
                continue;
            }
            if seen.len() > header.cells_count {
                return Err(Error("boc-unreachable-cells"));
            }
            for n in 0..cell.references_count() {
                pending.push(native(cell.reference(n))?);
            }
        }
        if seen.len() != header.cells_count {
            return Err(Error("boc-unreachable-cells"));
        }
    }
    Ok(root)
}
pub fn serialize_bytes(raw: &[u8]) -> Result<Vec<u8>, Error> {
    let bytes = native(chain_block::write_boc(&pack_bytes(raw)?))?;
    if bytes.len() > MAX_BOC {
        return Err(Error("boc-bound"));
    }
    Ok(bytes)
}
pub fn deserialize_bytes(bytes: &[u8], remaining: usize) -> Result<Vec<u8>, Error> {
    unpack_bytes(read_boc(bytes, false)?, remaining)
}
